packages/appeals-service-api/src/routes/v2/appeal-cases/repo.js

```diff
@@ -39,17 +82,142 @@ class AppealCaseRepository {
 	 */
-	putByCaseReference(caseReference, data) {
-		return this.dbClient.appealCase.upsert({
-			create: { ...data, Appeal: { create: {} } }, // create with parent Appeal
-			update: data,
+	async putHASByCaseReference(caseReference, data) {
+		const mappedData = mapHASDataModelToAppealCase(data);
+
+		// todo: move this to mapping
+		delete mappedData.caseType;
+		delete mappedData.linkedCaseStatus;
+		delete mappedData.leadCaseReference;
+		delete mappedData.notificationMethod;
+		delete mappedData.nearbyCaseReferences;
+		delete mappedData.neighbouringSiteAddresses;
+		delete mappedData.affectedListedBuildingNumbers;
+
+		const appealCase = await this.dbClient.appealCase.upsert({
+			create: { ...mappedData, Appeal: { create: {} } },
+			update: mappedData,
 			where: {
 				caseReference
 			}
 		});
+
+		// case relations
+		await this.dbClient.$transaction(async (tx) => {
+			await tx.appealCaseRelationship.deleteMany({
+				where: {
+					OR: [
+						{
+							caseReference: caseReference
+						},
+						{
+							caseReference2: caseReference
+						}
+					]
+				}
+			});
+
+			if (data.nearbyCaseReferences?.length) {
+				const direction1 = data.nearbyCaseReferences.map((nearby) => ({
+					caseReference: caseReference,
+					caseReference2: nearby
+				}));
+				const direction2 = data.nearbyCaseReferences.map((nearby) => ({
+					caseReference: nearby,
+					caseReference2: caseReference
+				}));
+
+				await tx.appealCaseRelationship.createMany({
+					data: direction1.concat(direction2)
+				});
+			}
+
+			if (data.leadCaseReference) {
+				await tx.appealCaseRelationship.create({
+					data: {
+						caseReference,
+						caseReference2: data.leadCaseReference,
+						type: 'linked' // todo: const this
+					}
+				});
+			}
+		});
+
+		// neighbour addresses
+		await this.dbClient.$transaction(async (tx) => {
+			await tx.neighbouringAddress.deleteMany({
+				where: {
+					caseReference: caseReference
+				}
+			});
+
+			if (data.neighbouringSiteAddresses?.length) {
+				await tx.neighbouringAddress.createMany({
+					data: data.neighbouringSiteAddresses.map((address) => ({
+						caseReference,
+						addressLine1: address.neighbouringSiteAddressLine1,
+						addressLine2: address.neighbouringSiteAddressLine2,
+						townCity: address.neighbouringSiteAddressTown,
+						county: address.neighbouringSiteAddressCounty,
+						postcode: address.neighbouringSiteAddressPostcode,
+						siteAccessDetails: address.neighbouringSiteAccessDetails,
+						siteSafetyDetails: address.neighbouringSiteSafetyDetails
+					}))
+				});
+			}
+		});
+
+		// listed building
+		await this.dbClient.$transaction(async (tx) => {
+			await tx.appealCaseListedBuilding.deleteMany({
+				where: {
+					caseReference: caseReference
+				}
+			});
+
+			if (data.affectedListedBuildingNumbers?.length) {
+				// todo: inefficient promise.all?
+				data.affectedListedBuildingNumbers.forEach(async (reference) => {
+					await tx.listedBuilding.upsert({
```

Comment:
be careful here this won't actually be awaited. Try
```
const randomTime = (id) => new Promise(res => {
    const waitTime = Math.random() * 1000
    setTimeout(() => res(`Promise ${id} waited ${waitTime}ms`), waitTime)
})

const 
outerFunc = async () => {
    await (new Array(5).fill(0)).forEach(async (_,id) => {const res = await randomTime(id); console.log(res)})
    console.log('outerfunc exited')
}

await outerFunc()
```
result
<img width="310" alt="Screenshot 2024-07-01 at 13 42 57" src="https://github.com/Planning-Inspectorate/appeal-planning-decision/assets/35689158/4ba8aeb3-e15b-43de-8126-b64a361351b8">

Reply:
`forEach` -> `map` + Promise.all ?

Reply:
thanks both, have fixed/have a test case